### PR TITLE
Refactor Synapse plugins, add matrix-synapse-pam

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Avoid having the core server depend on plugins. Also introduce a new one.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by Ralith** [comment]
Haven't had time to test, but seems like a good idea.

File: nixos/modules/services/misc/matrix-synapse.nix [hunk 1]
```diff
@@ -118,6 +121,14 @@ in {
           Overridable attribute of the matrix synapse server package to use.
         '';
       };
+      plugins = mkOption {
+        type = types.listOf types.package;
+        default = with pkgs.matrix-synapse-plugins; [ matrix-synapse-ldap3 matrix-synapse-pam ];
+        defaultText = "with pkgs.matrix-synapse-plugins [ matrix-synapse-ldap3 matrix-synapse-pam ]";
```
Comment (pacien) [inline nixos/modules/services/misc/matrix-synapse.nix]: Maybe those plugins shouldn't be enabled by default, in particular the newly introduced one.

File: pkgs/servers/matrix-synapse/default.nix [hunk 2]
```diff
@@ -1,27 +1,13 @@
-{ lib, stdenv, python3, openssl
-, enableSystemd ? stdenv.isLinux
+{ stdenv, lib, python, buildPythonApplication, fetchPypi
```
Comment (pacien) [inline pkgs/servers/matrix-synapse/default.nix]: `python` seems to be an alias for `python2`, not `python3`.
https://github.com/NixOS/nixpkgs/blob/daac44f/pkgs/top-level/all-packages.nix#L8873

Comment (pacien) [reply pkgs/servers/matrix-synapse/default.nix]: Nevermind, just noticed that the package was called with `python3.pkgs.callPackage` in `top-level/all-packages.nix`.

**Review by pacien — APPROVED** [review]

File: nixos/modules/services/misc/matrix-synapse.nix [hunk 3]
```diff
@@ -118,6 +121,14 @@ in {
           Overridable attribute of the matrix synapse server package to use.
         '';
       };
+      plugins = mkOption {
+        type = types.listOf types.package;
+        default = with pkgs.matrix-synapse-plugins; [ matrix-synapse-ldap3 matrix-synapse-pam ];
```
Comment (zimbatm) [inline nixos/modules/services/misc/matrix-synapse.nix:126]:
```suggestion
        default = with pkgs.matrix-synapse-plugins; [ ];
```

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -4327,7 +4327,8 @@ in
 
   makebootfat = callPackage ../tools/misc/makebootfat { };
 
-  matrix-synapse = callPackage ../servers/matrix-synapse { };
+  matrix-synapse = python3.pkgs.callPackage ../servers/matrix-synapse { };
+  matrix-synapse-plugins = matrix-synapse.python.pkgs.callPackage ../servers/matrix-synapse/plugins { };
```
Comment (zimbatm) [inline pkgs/top-level/all-packages.nix:4331]:
```suggestion
  matrix-synapse-plugins = recurseIntoAttrs (matrix-synapse.python.pkgs.callPackage ../servers/matrix-synapse/plugins { });
```

**Review by zimbatm — COMMENTED** [review]
:+1: good idea

**Comment by zimbatm** [comment]
cleaned-up the PR a bit. It should be good to go.

**Review by pacien — APPROVED** [review]

**Comment by ajs124** [comment]
There's a typo in `plugins.exampleText`.

`"with config.services.matrix-synapse.package.plugins [ matrix-synapse-ldap3 matrix-synapse-pam ]";`
should be
`"with config.services.matrix-synapse.package.plugins; [ matrix-synapse-ldap3 matrix-synapse-pam ]";`

**Pull request merged**